### DISCUSSION TITLE
Update user API giving unauthorised error

### DISCUSSION THREAD

**Discussion by snehalsabade** [Sunbird-Lern/Community#164] [Issues]
Hello @AmiableAnil ,

**Sunbird Version - 4.10.0**
I follow below link-
https://lern.sunbird.org/learn/product-and-developer-guide/user-and-org-service/apis/user-management

I am using user update API to update user organisation, In that I am trying v1/v2/v3 but getting not authorized error.

**Request** -
Request URL: https://host/api/user/v1/update
Request Method: PATCH
Status Code: 401 Not authorized

**Payload**- 
    "request": {
        "organisationId": "0136188527577661441067",
        "userId": "9a1ae74c-4035-4bf7-ab8c-7738231d8f50",
        "roles": [
            "CONTENT_CREATOR","PUBLIC"
        ],
        "username": "shivchanneltest_mdjw"
    }

**Response** -
{
    "id": "api.user.update",
    "ver": "v1",
    "ts": "2023-10-06 06:22:17:793+0000",
    "params": {
        "resmsgid": "cdfb205f9cb646be2cfeab064ca3caa4",
        "msgid": "cdfb205f9cb646be2cfeab064ca3caa4",
        "err": "UOS_USRUPDUOS_0070",
        "status": "FAILED",
        "errmsg": "You are not authorized."
    },
    "responseCode": "UNAUTHORIZED",
    "result": {}
}
I want to update the user organisation.
Please anyone can help me in this? 

cc  @komalm , @paritshivani , @Renuka-S

**Comment by AmiableAnil** [comment]
@snehalsabade @komalm @paritshivani , @Renuka-S Please user the below curl to add/remove a user to any org.
`curl --location 'https://staging.sunbirded.org/learner/user/v2/role/assign' \
--header 'Content-Type: text/plain' \
--data '{
    "request": {
        "userId": "56c84dee-7149-47af-902d-0138e080cec0",
        "roles": [
            {
                "role": "PUBLIC",
                "operation": "add",
                "scope": [
                    {
                        "organisationId": "01269878797503692810"
                    }
                ]
            }
        ]
    }
}'`

**Reply by Jayaprakash8887** [reply]
@snehalsabade Kindly confirm if the query is resolved and there are no further queries. If yes, kindly mark the thread as answered

**Comment by snehalsabade** [comment]
Hello @AmiableAnil,
Above issue is not resolved, I checked with {{host}}/user/v2/role/assign I am getting success response but user organization is not get update.
**Request**- 
{
    "request": {
        "userId": "7bfcd45f-7cc6-44c1-96b6-660fa853fb1b",
        "roles": [
            {
                "role": "PUBLIC",
                "operation": "add",
                "scope": [
                    {
                        "organisationId": "0130152125009674241"
                    }
                ]
            }
        ]
    }
}

{{host}}user/v1/read/7bfcd45f-7cc6-44c1-96b6-660fa853fb1b
**Response** - 
{
    "id": "api.user.read.7bfcd45f-7cc6-44c1-96b6-660fa853fb1b",
    "ver": "v1",
    "ts": "2023-10-09 05:17:36:041+0000",
    "params": {
        "resmsgid": "44b88cba946dcd21ba489d661980d4b5",
        "msgid": "44b88cba946dcd21ba489d661980d4b5",
        "err": null,
        "status": "SUCCESS",
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "response": {
            "webPages": null,
            "maskedPhone": "******9009",
            "tcStatus": null,
            "loginId": null,
            "subject": null,
            "channel": "niua",
            "profileUserTypes": [],
            "language": null,
            "updatedDate": null,
            "password": null,
            "managedBy": null,
            "flagsValue": 4,
            "id": "7bfcd45f-7cc6-44c1-96b6-660fa853fb1b",
            "recoveryEmail": "",
            "identifier": "7bfcd45f-7cc6-44c1-96b6-660fa853fb1b",
            "thumbnail": null,
            "profileVisibility": null,
            "updatedBy": null,
            "accesscode": null,
            "locationIds": [],
            "registryId": null,
            "rootOrgId": "0130701891041689600",
            "prevUsedEmail": "",
            "firstName": "snehaltest2",
            "profileLocation": [],
            "tncAcceptedOn": null,
            "allTncAccepted": {},
            "profileDetails": null,
            "phone": "******9009",
            "dob": null,
            "grade": null,
            "currentLoginTime": null,
            "userType": null,
            "status": 1,
            "lastName": "",
            "tncLatestVersion": "v12",
            "aadhaarno": null,
            "gender": null,
            "prevUsedPhone": "",
            "stateValidated": true,
            "isDeleted": false,
            "organisations": [
                {
                    "isSelfDeclaration": false,
                    "organisationId": "0130701891041689600",
                    "updatedBy": null,
                    "addedByName": null,
                    "addedBy": null,
                    "associationType": 1,
                    "approvedBy": null,
                    "updatedDate": null,
                    "userId": "7bfcd45f-7cc6-44c1-96b6-660fa853fb1b",
                    "approvaldate": null,
                    "isSystemUpload": false,
                    "isDeleted": false,
                    "hashTagId": "0130701891041689600",
                    "isSSO": true,
                    "isRejected": null,
                    "id": "0138934329324257281",
                    "position": null,
                    "isApproved": null,
                    "orgjoindate": "2023-09-29 08:05:44:787+0000",
                    "orgLeftDate": null
                }
            ],
            "provider": null,
            "countryCode": null,
            "tncLatestVersionUrl": "https://nuisstaging.blob.core.windows.net/termsandcondtions/terms-and-conditions-v12.html",
            "maskedEmail": "sn*********@yopmail.com",
            "regorgid": null,
            "tempPassword": null,
            "email": "sn*********@yopmail.com",
            "rootOrg": {
                "dateTime": null,
                "preferredLanguage": null,
                "keys": {},
                "organisationSubType": null,
                "approvedBy": null,
                "channel": "niua",
                "description": "NIUA-123456789",
                "updatedDate": "2023-10-04 15:29:18:961+0000",
                "addressId": null,
                "organisationType": 5,
                "orgType": null,
                "isTenant": true,
                "provider": null,
                "locationId": null,
                "orgCode": null,
                "theme": null,
                "id": "0130701891041689600",
                "isApproved": null,
                "communityId": null,
                "isBoard": true,
                "email": null,
                "slug": "niua",
                "isSSOEnabled": null,
                "thumbnail": null,
                "orgName": "NIUA",
                "updatedBy": "73edf1b6-4cd2-457c-a121-1dda7a638248",
                "locationIds": [],
                "externalId": null,
                "orgLocation": [],
                "isRootOrg": true,
                "rootOrgId": "0130701891041689600",
                "approvedDate": null,
                "imgUrl": null,
                "homeUrl": null,
                "orgTypeId": null,
                "isDefault": null,
                "createdDate": "2020-07-23 05:26:38:020+0000",
                "createdBy": null,
                "parentOrgId": null,
                "hashTagId": "0130701891041689600",
                "noOfMembers": null,
                "status": null
            },
            "phoneVerified": true,
            "profileSummary": null,
            "tcUpdatedDate": null,
            "recoveryPhone": "",
            "avatar": null,
            "userName": "username_1695974744706",
            "userId": "7bfcd45f-7cc6-44c1-96b6-660fa853fb1b",
            "userSubType": null,
            "promptTnC": true,
            "emailVerified": true,
            "lastLoginTime": null,
            "createdDate": "2023-09-29 08:05:44:775+0000",
            "framework": {},
            "createdBy": "20431439-c03e-4e3d-af30-e0fe38768cde",
            "profileUserType": {},
            "location": null,
            "tncAcceptedVersion": null
        }
    }
}
Can you please look into it.
cc @komalm , @paritshivani , @Renuka-S

**Comment by Jayaprakash8887** [comment]
@snehalsabade PUBLIC is a default role in the system. Please try with a different role. Entry of the same can be found in 'sunbird.user_roles' table in cassandra.

**Comment by snehalsabade** [comment]
Hello @AmiableAnil ,
- Above issue not resolve, I tried with your given request.
- I had tried with learner/user/v2/role/assign API but getting error.
- I want to select other organisation, while updating organisation from dropdown getting errror, when click on submit getting error.
Please can you look into it.

Request - 
	{
    "userId": "ba199247-4d87-4b95-a33d-153eaeaffd09",
    "roles": [
        {
            "role": "PUBLIC",
            "operation": "add",
            "scope": [
                {
                    "organisationId": "0130701891041689600"
                }
            ]
        },
        {
            "role": "ORG_ADMIN",
            "operation": "add",
            "scope": [
                {
                    "organisationId": "0130701891041689600"
                }
            ]
        },
        {
            "role": "CONTENT_CREATOR",
            "operation": "add",
            "scope": [
                {
                    "organisationId": "0130701891041689600"
                }
            ]
        }
    ]
}

curl '{{host}}/learner/user/v2/role/assign' \
  -X 'PATCH' \
  -H 'Accept: application/json' \
  -H 'Accept-Language: en-GB,en;q=0.9' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: connect.sid=s%3ACbxFr9dIVnMSWAEsidWnJwvu_gusOn2o.Qwf53KkujiyFiycKT30rhQ4Y4YHuVVrAFDUVqcgh0Ek' \
  -H 'Origin: http://localhost:3000' \
  -H 'Referer: http://localhost:3000/dashBoard/addusess' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36' \
  -H 'X-App-Id: local.sunbird.portal' \
  -H 'X-App-Version: 4.10.0' \
  -H 'X-Channel-Id: 0130701891041689600' \
  -H 'X-Device-ID: c8693ead0bf871dcb459be85636c3f82' \
  -H 'X-Org-code: 0130701891041689600' \
  -H 'X-Request-ID: f31e72ea-dccc-31ed-e6aa-ba6e0a04ace0' \
  -H 'X-Session-ID: CbxFr9dIVnMSWAEsidWnJwvu_gusOn2o' \
  -H 'X-Source: web' \
  -H 'X-User-ID: 20431439-c03e-4e3d-af30-e0fe38768cde' \
  -H 'X-msgid: f31e72ea-dccc-31ed-e6aa-ba6e0a04ace0' \
  -H 'sec-ch-ua: "Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Linux"' \
  -H 'ts: 2023-10-10T13:12:04+05:30' \
  --data-raw '{"request":{"userId":"ba199247-4d87-4b95-a33d-153eaeaffd09","roles":[{"role":"PUBLIC","operation":"add","scope":[{"organisationId":"0130701891041689600"}]},{"role":"ORG_ADMIN","operation":"add","scope":[{"organisationId":"0130701891041689600"}]},{"role":"CONTENT_CREATOR","operation":"add","scope":[{"organisationId":"0130701891041689600"}]}]}}' \
  --compressed
  
  Response - 
  For request 'PATCH /v2/user/assign/role'  (Not Found)
  
Case 2-
 When admin select role for user is only PUBLIC
  
  curl '{{host}}/learner/user/v2/role/assign' \
  -X 'PATCH' \
  -H 'Accept: application/json' \
  -H 'Accept-Language: en-GB,en;q=0.9' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: connect.sid=s%3AQ_7FTGOs2_6KHJxONQJxW4CrHK1qeEN1.y83lZM8Fnvdl9ZRB1xw%2FQuaUHp1bdcUhF4%2FG%2BnGaOq4' \
  -H 'Origin: http://localhost:3000' \
  -H 'Referer: http://localhost:3000/dashBoard/addusess' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36' \
  -H 'X-App-Id: local.sunbird.portal' \
  -H 'X-App-Version: 4.10.0' \
  -H 'X-Channel-Id: 0130701891041689600' \
  -H 'X-Device-ID: c8693ead0bf871dcb459be85636c3f82' \
  -H 'X-Org-code: 0130701891041689600' \
  -H 'X-Request-ID: 819431e6-89ab-bbde-6666-afb3fceae101' \
  -H 'X-Session-ID: Q_7FTGOs2_6KHJxONQJxW4CrHK1qeEN1' \
  -H 'X-Source: web' \
  -H 'X-User-ID: 20431439-c03e-4e3d-af30-e0fe38768cde' \
  -H 'X-msgid: 819431e6-89ab-bbde-6666-afb3fceae101' \
  -H 'sec-ch-ua: "Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Linux"' \
  -H 'ts: 2023-10-10T13:16:09+05:30' \
  --data-raw '{"request":{"userId":"ba199247-4d87-4b95-a33d-153eaeaffd09","roles":[{"role":"PUBLIC","operation":"add","scope":[{"organisationId":"0130701891041689600"}]}]}}' \
  --compressed
  
  Request- 
  {
    "userId": "ba199247-4d87-4b95-a33d-153eaeaffd09",
    "roles": [
        {
            "role": "PUBLIC",
            "operation": "add",
            "scope": [
                {
                    "organisationId": "0130701891041689600"
                }
            ]
        }
    ]
}

 Response - 
  For request 'PATCH /v2/user/assign/role'  (Not Found)
  cc @komalm , @paritshivani , @Renuka-S

**Reply by BharathwajShankar** [reply]
@snehalsabade v2/user/assign/role is a post API and please find a sample curl here,

curl --location {{host}}/api/user/v2/role/assign' \
--header 'Authorization: {{auth_token}}' \
--header 'X-Authenticated-User-token: {{keycloak_token}}' \
--header 'Content-Type: application/json' \
--data '{
    "request": {
        "userId": "{{user_id}",
        "roles": [
            {
                "role": "CONTENT_CREATOR",
                "operation": "add",
                "scope": [
                    {
                        "organisationId": "{{org_id}}"
                    }
                ]
            }
        ]
    }
}'

**Comment by BharathwajShankar** [comment]
@snehalsabade two observations here, 

1. If the user update API is giving an unauthorized error make sure you are using the same auth token as the user ID you are trying to update. for example, if you are using the user id of 9a1ae74c-4035-4bf7-ab8c-7738231d8f50 then you should use the username and password of this particular user. to trigger the update API
2. role assign API can be used to update the roles of the particular user, if your intention is to update the details of this particular user then check the 1st point and use it accordingly it should help you.

**Comment by snehalsabade** [comment]
Hello @BharathwajShankar,
- using **learner/user/v2/role/assign** this able to edit the role of particular user. 
- as an org-admin role want to change **organisation** of the user.
Please can you help me for that.

cc @komalm , @paritshivani , @Renuka-S

**Reply by BharathwajShankar** [reply]
@snehalsabade it is not possible to update the user's organisation with the APIs available. We have a user migrate API where you can migrate to  non-custodian organisation from custodian organisation.

**Reply by snehalsabade** [reply]
Hello @BharathwajShankar ,
In my case the user is in root organisation and I want to add sub-organisation to the user with roles.
If you are available Please can I schedule a meeting?
cc @komalm , @paritshivani , @Renuka-S

**Reply by snehalsabade** [reply]
Hello @BharathwajShankar ,
In my case the user is in root organisation and I want to add sub-organisation to the user with roles.
In that case which API I have to use ?
cc @komalm , @paritshivani , @Renuka-S

**Reply by BharathwajShankar** [reply]
@snehalsabade please use user migrate API in this regard, sample curl for the same

curl --location --request PATCH '{{host}}/api/user/v1/migrate' \
--header 'Authorization: "<kong_api_key>"' \
--header 'X-Authenticated-User-token: "<keycloak-token>"' \
--header 'Content-Type: application/json' \
--data '{
    "request": {
        "userId": "",
        "channel": "",
        "externalIds": [
            {
                "id": "",
                "idType": "",
                "provider": ""
            }
        ]
    }
}'

**Reply by snehalsabade** [reply]
Hello @BharathwajShankar ,
**case 1-**
- I am using PATCH '{{host}}/api/user/v1/migrate' API, In that what is the value for **idType**.
- by using this got 404 not found response.
**Request-** 
{
  "organisationId": ,
  "userId": ,
  "username": ,
  "roles": [
    {
      "role": "PUBLIC"
    },
  ],
  "externalIds": [
    {
      "idType": ,
      "provider": 
    }
  ]
}
**case 2-**
- https://lern.sunbird.org/use/developer-guide/user-and-org-service/apis/user-management  
- using this POST'{{host}}/api/user/v1/migrate api  got 404 not found response
- In that what is the value for feedId , userExtId?
 **Request** -
 {
  "request": {
    "userId": ,
    "channel": ,
    "action": ,
    "feedId": ,
    "userExtId": 
  }
}
Can you please guide me here.
cc @komalm , @paritshivani , @Renuka-S

**Reply by BharathwajShankar** [reply]
@snehalsabade for case 1: the value of idType is nothing but the org-id of the user. Also, the API endpoint is,
{{host}}/private/user/v1/migrate

Refer to below threads as well in addition on any queries,
https://github.com/orgs/Sunbird-Lern/discussions/104
https://github.com/orgs/Sunbird-Lern/discussions/51

**Reply by Jayaprakash8887** [reply]
@snehalsabade Please mark the thread as answered if there are no further queries.

**Reply by snehalsabade** [reply]
Hello @BharathwajShankar ,
- When I am using post/patch method both the methods giving 302 error.
- {
    "**request**": {
        "userId": "",
        "channel": [ ],
        "externalIds": [
            {
                "id": "",
                "idType": "",
                "provider": ""
            }
        ]
    }
}
 Can you please guide me here.
cc @komalm , @paritshivani , @Renuka-S

**Reply by gandham-santhosh** [reply]
@snehalsabade Its redirection. can you share the hostname you are using in your call? Also, please check if the request is passing NGINX and Kong by going through the logs

**Reply by snehalsabade** [reply]
Hello @Nithin-ns24
Issue is still open
can you update latest logs for it? 
cc @komalm , @paritshivani

**Comment by AmiableAnil** [comment]
Closing this thread. Please feel free to reopen if required.

**Comment by snehalsabade** [comment]
@gandham-santhosh 
Here I am sharing learner logs, Please find below attached file
[add-org lerner logs.docx](https://github.com/Sunbird-Lern/Community/files/14278067/add-org.lerner.logs.docx)

cc @komalm , @paritshivani,
